clang/lib/CodeGen/CGVTables.cpp

```diff
@@ -1180,6 +1185,21 @@ bool CodeGenVTables::isVTableExternal(const CXXRecordDecl *RD) {
       TSK == TSK_ExplicitInstantiationDefinition)
     return false;
 
+  // Itanium C++ ABI [5.2.3]:
+  // Virtual tables for dynamic classes are emitted as follows:
+  //
+  // - If the class is templated, the tables are emitted in every object that
+  // references any of them.
+  // - Otherwise, if the class is attached to a module, the tables are uniquely
+  // emitted in the object for the module unit in which it is defined.
+  // - Otherwise, if the class has a key function (see below), the tables are
+  // emitted in the object for the translation unit containing the definition of
+  // the key function. This is unique if the key function is not inline.
+  // - Otherwise, the tables are emitted in every object that references any of
+  // them.
+  if (Module *M = RD->getOwningModule(); M && M->isNamedModule())
```

Comment:
Rather than doing this based on a named module, I wonder if this could be powered by the same system that originally worked for clang header modular code generation. It worked/works by adding a bit to declarations about whether to home them to a modular object file

e6b7c28d17edf428c58ba65c549109c0a488a5be

Though I guess in 1ac9c98e6c40d9a1e6cb904b414161388af4a89c I used the bit on a type to mean "home this type in debug info" and the bit on a function to mean "home this function's code" - so I'm not sure what we'd do to track the vtable specifically... Maybe not worth it/not feasible, not sure.

Reply:
I guess doing it through the `hasExternalDefinitions` system suggested above might reduce the number of special cases here, if there was a nice way to do address the code V debug info use.

Reply:
Maybe it is not strictly correct to use `hasExternalDefinitions()` since it can't handle the global module case. I mean, from the current wording, the class in the global module should follow the old ABI rules (key functions). But `hasExternalDefinitions()`  won't make that clear.

Reply:
`hasExternalDefinitions` isn't "is this AST from some other AST file" but "did the AST file say it'd handle the object-file definition of this thing" - the pcm file specifies which things it'll handle.

So if the intent is not to home anything in the GMF, then pcms could be written out that don't set "hasExternalDefinitions" to true for those entities.

Reply:
> hasExternalDefinitions isn't "is this AST from some other AST file" but "did the AST file say it'd handle the object-file definition of this thing" - the pcm file specifies which things it'll handle.
> So if the intent is not to home anything in the GMF, then pcms could be written out that don't set "hasExternalDefinitions" to true for those entities.

No, it is not the intent. The intent is to keep GMF as legacy code instead of named modules.

For example,

```
// header.h
void func() { } // ! Non inline

// x.cppm
module;
#include "header.h"
export module x;
...
```

Then in the object file of `x.cppm`, (if `func()` is not discarded out), we should be able to see the definition of `func()`. Although it is rare since the definitions in the headers are generally inline, but we as the compiler can't assume that.

I think the current implementation is more straight forward and clear.

Reply:
Even if we have to rename "has external definitions" to something else to generalize the concept, I still think it would be nice to avoid having multiple ways to determine whether something is "homed" to the modular object file... - though the difference between the default/obvious behavior for a linkage-having function (like `func()` above) in clang header modules (by default, goes everywhere the module is imported, probably creating an ODR violation) and C++ Modules is significant (if it were just "obvious" things it might be fine - but there's lots of different kinds of things and I worry trying to encode all of those through "has external definitions" might miss some, so I can understand having a different default for modules)

But it still seems like for cases where we want to override the default rules (non-inline definitions go wherever they're defined) then using the "has external definitions" mechanism might be good - again, to keep it consistent between different modules systems. And means fewer ad-hoc checks littered around the place that have to be kept in sync - emit the "has external definitions" bit when the module is created, it gets checked for when producing a definition, and that's all. (still needs some work here, since it currently only supports functions for code and types for debug info)

Reply:
Yeah, I agree it might be good conceptly. But given the clang header modules with object files are not widely used and the current patch may be somewhat impactful, I feel better to do that in the future when we really need to.

Reply:
The plumbing's already there though - and means one source of truth that both the consumer of the .pcm and the compilation of the .pcm->.o step use - rather than two separate pieces of code that also have to be kept in sync. (and the code's already there, the check before the one this patch proposes to be added to `isConsumerInterestedIn` is the modular code generation codepath already - and the populating of `ModularCodegenDecls` is part of that system)

I'd rather not add ad-hoc hardcoded things across these several places when we have a mechanism to use that gets set at .pcm time, is (sort of) generic across that for the usage and the implementation compilation.

It seems like fine existing infrastructure to further generalize to handle home-able types (that can be homed for code generation reasons, as well as for debug info reasons).

Reply:
Done. I add the interfaces in precommit: https://github.com/llvm/llvm-project/commit/99873b35da7ecb905143c8a6b8deca4d4416f1a9